### ISSUE TITLE
GetSchema Async

### ISSUE THREAD

**Issue by tanbchr** [dotnet/corefx#37940]
I was curious to know why there is not an `async` version of the `GetSchema`?

Basically, for each collection information name, it is doing queries to retrieve the metadata.

I know that we can query directly the database to retrieve the same information. However, we are using multiple ADO.NET providers and we are not depending on the implementation of the `GetSchema` function nor on the type of the connection.

G getting the metadata is really heavy and requires a lot of time to fetch everything. By having an asynchronous version, we can free the current thread.

**Comment by tanbchr** [comment]
We are about to wrap it with a TaskCompletionSource, so we can have an async version of it. But we are not 100% sure if it is a good idea. Maybe some concrete implementation of `GetSchema` are not async (I/O) and that is why there is not an asynchronous version of it?

**Comment by roji** [comment]
tl;dr it could make sense to add `GetSchemaAsync()` to ADO.NET, but this is likely a low-priority async addition.

More in detail:

Wrapping a synchronous method with TaskCompletionSource will not turn it into a "real" async method: you will still need some thread to block while the method is executing (see [this article](https://devblogs.microsoft.com/pfxteam/should-i-expose-asynchronous-wrappers-for-synchronous-methods/) even though it's targeted more at library writers). To have an actual async GetSchema(), each database provider would need to implement a `GetSchemaAsync()` using only async I/O internally.

In general it does make sense to provide async counterparts to all I/O-based methods - and GetSchema() is a likely candidate. There is indeed a wave of async APIs making it into ADO.NET (see #35012), but we first concentrated on more meaningful, heavily-used methods first.

One specific reason that I suspect GetSchemaAsync() would be less valuable, is that its results are in general expected to be cachable in the application. In other words, assuming more or less the same queries get executed in your application, you'd be better off eliminating I/O altogether by caching the results, rather than fetching those results over and over asynchronously. This won't hold in all cases but does explain why I believe it's not as important as other APIs.

**Comment by tanbchr** [comment]
I do understand that this can be a low priority and that it is not the most used method in the ADO.NET. However, for ETL processes, it is super important. 

Right now, we are making a homemade data-synchronizer that enables the users to select the tables that they want to migrate to other systems. The `GetSchema` is heavily used to fetch the metadata of the source connections when it is the time to synchronize. We already have a caching system that expires after an amount of time. The problem with caching is when the metadata get changed. It is difficult to exactly know when to flush the cache and we can have some consistency issues.

By delaying the integration of such feature in the roadmap will delay the library writers to react to it and it will take even more time to have it those libraries.

**Comment by roji** [comment]
> The problem with caching is when the metadata get changed.

Are you referring to actual tables getting changed, in runtime, and therefore the cache getting out of sync? That seems somewhat extreme as schema changes typically also involve code changes, which means your application is restarting anyway... More detail here would be appreciated.

If you have a caching, then we're talking about optimizing for not blocking a thread in the rare case where the cache expires - right? Note that I'm not saying the feature is worthless, I'm just trying to understand your use-case and if there's something I've missed.

**Comment by tanbchr** [comment]
The synchronizer application that we are building is able to connect to any source of data. We are using those ADO.NET libraries from this [vendor](https://www.cdata.com/ado/). 

We have no control over the metadata of those sources. Since we are not dependent on the sources that we read, we don't have to restart the data-synchronizer itself. We act as a third party application and can connect to any source.

Let's say a source gets updated (ERP/System/etc) and its metadata gets changed. For example, the type of a column of a table got changed somehow from integer to decimal. Therefore, the cache will not have the right value, unless it is flushed.

And you are right. Caching and async are 2 different subjects. By having the cache, we do not have to call the sync version of getSchema and slow down our application, we have right away the metadata that we want.

Is this clear enough or would like a more in-depth explanation?

**Comment by roji** [comment]
Thanks for the additional info, it's clear.

Note that as I wrote above, even if we were to add an async GetSchema() API in corefx, for it to actually be helpful providers would have to be updated to override it and actually perform the I/O asynchronously - we have no control over that. Just to make sure it's understood.

**Comment by tanbchr** [comment]
Yes! It is exactly why I said that it will be nice to have it sooner. It will give more time to the provider's writer to integrate it in their libraries.
